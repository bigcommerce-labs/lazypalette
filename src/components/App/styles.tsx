import styled from 'styled-components';

export const StyledApp = styled.div`
    background: ${({ theme: { colors } }) => colors.background};
    align-content: flex-start;
    display: flex;
    flex-direction: column;
    height: 100%;
    position: fixed;
    width: 100%;
`;

export const Viewport = styled.div`
    display: flex;
    flex-direction: row;
    flex-grow: 1;
`;
