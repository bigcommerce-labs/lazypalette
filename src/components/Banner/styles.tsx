import styled from 'styled-components';

export const StyledBanner = styled.div`
    background-color: #FFD24D;
    text-align: center;
    width: 100%;
    line-height: 2.5rem;
    font-size: 1rem;
    font-weight: 600;
    color: #303540;
`;

export const StyledIcon = styled.span`
    span {
        margin-bottom: -5px;
    }
`;

export const StyledMessage = styled.p`
    text-align: left;

    a {
        color: ${({ theme }) => theme.colors.primary};
        display: block;
        font-weight: bold;
        margin-top: 5px;
        text-decoration: none;
    }
`;

export const StyledShareableLink = styled.button`
    background: ${({ theme }) => theme.colors.warning};
    color: ${({ theme }) => theme.colors.primaryText};
    font-size: ${({ theme }) => theme.typography.fontSize.small};
    font-weight: ${({ theme }) => theme.typography.fontWeight.bold}
    font-family: ${({ theme }) => theme.typography.fontFamily.headings};
    margin: .5rem 0 .5rem 1.25rem;
    outline: none;
    border: 1px solid;
    border-radius: 2px;
    display: inline-block;
    padding: .25rem .5rem;
    line-height: 1.375rem;
    cursor: pointer;
`;
