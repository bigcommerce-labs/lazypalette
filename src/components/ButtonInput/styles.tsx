import styled from 'styled-components';

export const Wrapper = styled.div`
    margin: 0;
`;

interface ButtonProps {
    border?: boolean;
    classType?: string;
}

export const Button = styled.button.attrs<ButtonProps>({})`
    height: 2rem;
    border-radius: 2px;
    box-sizing: border-box;
    border: ${({ border, theme }) => border ? `1px solid ${theme.colors.primary}` : 'none'};
    padding: ${({ border }) => border ? '0 1rem' : '0'};
    font-family: ${({ theme }) => theme.typography.fontFamily.sans };
    font-size: 1rem;
    cursor: pointer;

    ${({ border, classType, theme }) => {
        const isPrimary = classType === 'primary';

        return `
            background-color: ${isPrimary ? theme.colors.primary : '#FFF'};
            color: ${isPrimary ? '#FFF' : theme.colors.primary};

            :hover {
                background-color: ${isPrimary ? theme.colors.primaryHover : theme.colors.secondaryHover};
                ${!border && `
                    background-color: transparent;
                `}
                border-color: ${isPrimary ? theme.colors.brandPrimary : theme.colors.primary};
                text-decoration: ${border ? 'none' : 'underline'};
            }

            :focus {
                outline: none !important;
            }

            :disabled {
                background-color: ${isPrimary ? theme.colors.primary : '#FFF'};
                opacity: 0.5;
            }
        `;
    }}
`;

Button.defaultProps = {
    theme: {
        colors: {
            background: '#F5F7FA',
            brandPrimary: '#273A8A',
            primary: '#3F69FE',
        },
        typography: {
            fontFamily: {
                sans: '"Source Sans Pro", "Helvetica Neue", Arial, sans-serif',
            },
        },
    },
};
