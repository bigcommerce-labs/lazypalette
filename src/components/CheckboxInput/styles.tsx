import styled from 'styled-components';

export const Container = styled.div`
    position: relative;
`;

export const Input = styled.input.attrs({
    type: 'checkbox',
})`
    ${({ theme }) => `
        opacity: 0;

        :not(:checked) + label:after {
            opacity: 0;
        }

        :not(:checked) + label {
            background: #FFF;
            border: 1px solid ${theme.colors.guideText};
            box-sizing: border-box;
        }
    `}
`;

export const Label = styled.label`
    ${({ theme }) => `
        cursor: pointer;
        position: absolute;
        width: 1.25rem;
        height: 1.25rem;
        top: 25%;
        left: 0;
        background: ${theme.colors.primary};
        border-radius: 2px;

        :after {
            content: '';
            position: absolute;
            width: .5rem;
            height: .25rem;
            background: transparent;
            top: .25rem;
            left: .25rem;
            border: 3px solid ${theme.colors.stroke};
            border-top: none;
            border-right: none;
            transform: rotate(-45deg);
        }
    `}
`;

export const HiddenLabel = styled.label`
    padding-left: .5rem;
    line-height: 1.25rem;
    display: inline-block;
`;
