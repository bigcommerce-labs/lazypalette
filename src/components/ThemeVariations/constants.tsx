export const enum Messages {
    Variation = 'This variation has unpublished changes. Do you want to proceed?',
}
