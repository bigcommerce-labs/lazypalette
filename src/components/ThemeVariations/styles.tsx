import styled from 'styled-components';

import { colors } from '../../styleConstants';

export const List = styled.ul`
  width: 56.25rem;
  height: 15.3125rem;
  list-style: none;
  padding: 0;
  margin: 0;
  display: flex;
  flex-wrap: wrap;
  color: #919596;
`;

export const Item = styled.li`
  margin: .625rem 1.5625rem;
  cursor: pointer;
`;

export const Title = styled.div`
  color: ${colors.midGray};
  font-weight: 400;
`;

export const Thumb = styled.div`
  width: 15.625rem;
  height: 10.9375rem;
  margin-top: .625rem;
  background: url("${props => props.theme}") left top/cover no-repeat;
`;
