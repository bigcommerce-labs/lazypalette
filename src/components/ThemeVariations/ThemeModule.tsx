import React from 'react';

import { Item, List, Thumb, Title } from './styles';
import { ThemePropsList } from './ThemeVariations';

const ThemeModule = (props: {variants: ThemePropsList}) => (
    <List>
        {props.variants.map(({name, image, isActive}) => (
            <Item key={name} isActive={isActive}>
                <Title>{name}</Title>
                <Thumb previewPath={image} />
            </Item>
        ))}
    </List>
);

export default ThemeModule;
