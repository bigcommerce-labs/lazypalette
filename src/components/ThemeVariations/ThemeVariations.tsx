import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { withRouter, Route, RouteComponentProps } from 'react-router-dom';
import { Dispatch } from 'redux';

import { loadTheme } from '../../actions/theme';
import { State } from '../../reducers/reducers';
import ExpandableMenu from '../ExpandableMenu/ExpandableMenu';

import Draggable from '../Draggable/Draggable';
import ConfirmModal from '../Modal/ConfirmModal/ConfirmModal';
import { appRoutes } from '../Routes/Routes';

import { Messages } from './constants';
import ThemeModule from './ThemeModule';

interface ThemeVariationsProps extends RouteComponentProps<{}> {
    isChanged: boolean;
    position: { x: number, y: number };
    themeVariants: ThemePropsList;
    loadTheme(variationID: string): Dispatch<State>;
}

export interface ThemePropsList extends Array<StateProps> {}

interface StateProps {
    isActive: boolean;
    image: string;
    name: string;
    variationId: string;
}

interface ThemeVariationsState {
    isConfirmOpen: boolean;
    variationId: string;
}

export class ThemeVariations extends PureComponent <ThemeVariationsProps, ThemeVariationsState> {
    readonly state: ThemeVariationsState = {
        isConfirmOpen: false,
        variationId: '',
    };

    open = (variationId: string) => this.setState({
        isConfirmOpen: true,
        variationId,
    });

    close = () => this.setState({
        isConfirmOpen: false,
        variationId: '',
    });

    handleChange = () => {
        const { variationId } = this.state;

        this.setState({
            isConfirmOpen: false,
            variationId: '',
        }, () => {
            this.props.loadTheme(variationId);
        });
    };

    handleVariationChange = (variationId: string): void => {
        const { isChanged } = this.props;

        isChanged ? this.open(variationId) : this.props.loadTheme(variationId);
    };

    render() {
        const { styles } = appRoutes;
        const { position, match, themeVariants} = this.props;
        const { isConfirmOpen } = this.state;

        return (
            <Route
                path={styles.path}
                exact
                render={() => (
                    <>
                        <Draggable position={position}>
                            <ExpandableMenu title="Styles" back={match.url}>
                                <ThemeModule
                                    variants={themeVariants}
                                    handleVariationChange={this.handleVariationChange}
                                />
                            </ExpandableMenu>
                        </Draggable>
                        {isConfirmOpen &&
                            <ConfirmModal
                                body={Messages.Variation}
                                primaryAction={this.close}
                                secondaryAction={this.handleChange}
                                title="Theme Change Warning"
                            />
                        }
                    </>
                )}
            />
        );
    }
}

const mapStateToProps = (state: State) => ({
    isChanged: state.theme.isChanged,
    themeVariants: state.theme.variations.map(({ screenshot, variationName, id }): StateProps => ({
        image: screenshot.smallThumb,
        isActive: state.theme.variationId === id,
        name: variationName,
        variationId: id,
    })),
});

const mapDispatchToProps = {
    loadTheme,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ThemeVariations));
