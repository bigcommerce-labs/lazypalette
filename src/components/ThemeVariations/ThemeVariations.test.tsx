import { mount, shallow } from 'enzyme';
import React from 'react';
import { StaticRouter } from 'react-router';

import { appRoutes } from '../Routes/Routes';

import { Item } from './styles';
import { ThemeVariations } from './ThemeVariations';

describe('ThemeVariations', () => {
    const { styles } = appRoutes;
    const mockProp: any = jest.fn();
    const routeProps = {
        history: mockProp,
        location: mockProp,
        match: mockProp,
        staticContext: mockProp,
    };

    routeProps.match.url = './';

    const testItems = [
        {
            image: 'http://meow.wow.com/123.jpg',
            isActive: true,
            name: 'light',
            variationId: '123',
        },
        {
            image: 'http://meow.wow.com/123.jpg',
            isActive: false,
            name: 'dark',
            variationId: '234',
        },
    ];

    it('renders', () => {
        const mockHandler = jest.fn();

        const themeVariations = shallow(
            <StaticRouter location={styles.path} context={{}}>
                <ThemeVariations
                    isChanged={false}
                    position={{ x: 5, y: 10}}
                    themeVariants={testItems}
                    loadTheme={mockHandler}
                    {...routeProps}
                />
            </StaticRouter>
        );

        expect(themeVariations).toMatchSnapshot();
    });

    describe('handleVariationChange', () => {
        describe('when there are no unsaved theme changes', () => {
            it('should call loadTheme', () => {
                const mockLoadTheme = jest.fn();

                const wrapper = mount(
                    <StaticRouter location={styles.path} context={{}}>
                        <ThemeVariations
                            isChanged={false} // no theme changes
                            position={{ x: 5, y: 10}}
                            themeVariants={testItems}
                            loadTheme={mockLoadTheme}
                            {...routeProps}
                        />
                    </StaticRouter>
                );

                const themeVariations = wrapper.find(ThemeVariations).instance();
                themeVariations.setState({ isConfirmOpen: false });
                wrapper.update();

                const otherThemeVariant = wrapper.find(Item).last();
                otherThemeVariant.simulate('click');

                // should directly call loadTheme
                expect(mockLoadTheme).toHaveBeenCalledTimes(1);
                expect(themeVariations.state.isConfirmOpen).toEqual(false);
            });
        });

        describe('when there are theme changes', () => {
            it('should open the confirm modal', () => {
                const mockHandler = jest.fn();

                const wrapper = mount(
                    <StaticRouter location={styles.path} context={{}}>
                        <ThemeVariations
                            isChanged={true} // theme has unsaved changes
                            position={{ x: 5, y: 10}}
                            themeVariants={testItems}
                            loadTheme={mockHandler}
                            {...routeProps}
                        />
                    </StaticRouter>
                );

                const themeVariations = wrapper.find(ThemeVariations).instance();
                themeVariations.setState({ isConfirmOpen: false });
                wrapper.update();

                const otherThemeVariant = wrapper.find(Item).last();
                otherThemeVariant.simulate('click');

                // should open confirm modal - not directly call loadTheme
                expect(mockHandler).toHaveBeenCalledTimes(0);
                expect(themeVariations.state.isConfirmOpen).toEqual(true);
            });
        });
    });
});
