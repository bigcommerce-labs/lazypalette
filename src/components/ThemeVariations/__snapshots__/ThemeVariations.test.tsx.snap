// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeVariations renders 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/styles",
        "search": "",
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <ThemeVariations
    history={[MockFunction]}
    isChanged={false}
    loadTheme={[MockFunction]}
    location={[MockFunction]}
    match={[MockFunction]}
    position={
      Object {
        "x": 5,
        "y": 10,
      }
    }
    staticContext={[MockFunction]}
    themeVariants={
      Array [
        Object {
          "image": "http://meow.wow.com/123.jpg",
          "isActive": true,
          "name": "light",
          "variationId": "123",
        },
        Object {
          "image": "http://meow.wow.com/123.jpg",
          "isActive": false,
          "name": "dark",
          "variationId": "234",
        },
      ]
    }
  />
</Router>
`;
